Firestore/Swift/Tests/Codable/CodableDocumentTests.swift

```diff
@@ -0,0 +1,405 @@
+import FirebaseFirestore
+@testable import FirebaseFirestoreSwift
+import XCTest
+
+class CodableDocumentTests: XCTestCase {
+  func roundTrip<X>(input: X, expected: [String: Any]? = nil) -> X where X: Codable {
+    var encoded = [String: Any]()
+    do {
+      encoded = try Firestore.encode(input)
+      if let expected = expected {
+        XCTAssertEqual(encoded as NSDictionary, expected as NSDictionary)
+      }
+    } catch {
+      XCTFail("Failed to encode \(X.self): error: \(error)")
+    }
+    do {
+      let decoded = try Firestore.decode(X.self, from: encoded)
+      return decoded
+    } catch {
+      XCTFail("Failed to decode \(X.self): \(error)")
+    }
+    return input // After failure
+  }
+
+  func testInt() {
+    struct Model: Codable {
+      let x: Int
+    }
+    let model = Model(x: 42)
+    let dict = ["x": 42]
+    XCTAssertEqual(model.x, roundTrip(input: model, expected: dict).x)
+  }
+
+  func testEmpty() {
+    struct Model: Codable {}
+    let model = Model()
+    let dict = [String: Any]() as NSDictionary
+    let encoded = try! Firestore.encode(model) as NSDictionary
+    XCTAssertEqual(encoded, dict)
+  }
+
+  func testNil() {
+    struct Model: Codable {
+      let x: Int?
+    }
+    let model = Model(x: nil)
+    let dict = ["x": nil] as [String: Int?]
+    let encodedDict = try! Firestore.encode(model)
+    XCTAssertNil(encodedDict["x"])
+    let model2 = try? Firestore.decode(Model.self, from: dict as [String: Any])
+    XCTAssertNil(model2)
+  }
+
+  func testOptional() {
+    struct Model: Codable {
+      let x: Int
+      let opt: Int?
+    }
+    let dict = ["x": 42]
+    let model = Model(x: 42, opt: nil)
+    XCTAssertEqual(model.x, roundTrip(input: model, expected: dict).x)
+
+    let model2 = Model(x: 42, opt: 7)
+    let expected = ["x": 42, "opt": 7]
+    let encoded = try! Firestore.encode(model2)
+    XCTAssertEqual(encoded as NSDictionary, expected as NSDictionary)
+    let decoded = try! Firestore.decode(Model.self, from: expected)
+    XCTAssertEqual(decoded.x, model2.x)
+    XCTAssertEqual(decoded.opt, model2.opt)
+  }
+
+  func testOptionalTimestamp() {
+    struct Model: Codable {
+      let value: Int
+      let timestamp: Timestamp?
+    }
+    class FirestoreDummy {
+      var visited = 0
+      func setObject<T: Codable>(_ object: T, fieldValues: [PartialKeyPath<T>: FieldValue] = [:]) {
+        let obj = object as! Model
+        XCTAssertEqual(obj.value, 10)
+        XCTAssertNil(obj.timestamp)
+        visited += 1
+      }
+    }
+    let c = Model(value: 10, timestamp: nil)
+    let fs = FirestoreDummy()
+    // If no custom field values need to be set:
+    fs.setObject(c)
+
+    // Or, overriding custom field values:
+    fs.setObject(c, fieldValues: [\Model.timestamp: FieldValue.serverTimestamp(),
+                                  \Model.value: FieldValue.delete()])
+    XCTAssert(fs.visited == 2)
+  }
+
+  func testEnum() {
+    enum MyEnum: Codable, Equatable {
```

Comment:
As implemented here it appears this enumeration may be suitable for encoding both as a top-level type and when nested. Pull this out and assert both round-trips work?

Are enums ever directly Codable (i.e. without writing this boilerplate)? If so we should have a test for that too.

Reply:
Not that I could find.